File: 프로그래머스/lv2/송예은/n^2 배열 자르기.py
# 구현 문제 - left와 right에 해당되는 인덱스를 구한 다음 y와 x중 큰 값의 인덱스+1의 값이다.-> 시간 초과 해결못함

def solution(n, left, right):
    arr = [[0 for _ in range(n)] for _ in range(n)]
    newArr = []
    curr_num = -1
    isPossible = True
    
    for i in range(n):
        if not isPossible:
            break
        for j in range(n):
            arr[i][j] = max(i,j)+1
            arr[j][i] = max(i,j)+1
            curr_num += 1
            if curr_num > right:
                isPossible = False
                break
            if curr_num >= left and curr_num <=right:
                newArr.append(arr[i][j])
    return newArr

File: 프로그래머스/lv2/지준혁/n^2 배열 자르기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// 1 2 3
// 2 2 3
// 3 3 3
// -> 1 2 {3 2 2 3} 3 3 3
// 규칙을 찾을 수 있니?
// n = 3, 시작인덱스: 1
// 0 -> (0, 0)
// 1 -> (0, 1)
// 2 -> (0, 2)
// 3 -> (1, 0)
// 4 -> (1, 1)
// 5 -> (1, 2)
// 6 -> {2, 0}, {2, 1}, {2, 2}
//
// 규칙에 따라 숫자를 채워넣는 방식 최소 100만^2 시간초과
// 인덱스를 뽑아서 규칙 찾아내는 방식
vector<int> solution(int n, long long left, long long right) {
    
    vector<int> answer;
    for (long long i = left; i <= right; ++i) {
        long long share = i / n;
        long long remainder = i % n;
        answer.push_back(max(share, remainder) + 1);
    }
    return answer;
}
